File: bots/guardian/game.php
<?php
/**
 * Game state types for AI Code Battle protocol.
 */

/**
 * Position on the grid
 */
class Position {
    public int $row;
    public int $col;

    public function __construct(int $row, int $col) {
        $this->row = $row;
        $this->col = $col;
    }

    public static function fromArray(array $data): self {
        return new self($data['row'], $data['col']);
    }

    public function toArray(): array {
        return ['row' => $this->row, 'col' => $this->col];
    }

    /**
     * Move in a direction with toroidal wrapping
     */
    public function moveToward(string $dir, int $rows, int $cols): Position {
        switch ($dir) {
            case 'N':
                return new Position(($this->row - 1 + $rows) % $rows, $this->col);
            case 'E':
                return new Position($this->row, ($this->col + 1) % $cols);
            case 'S':
                return new Position(($this->row + 1) % $rows, $this->col);
            case 'W':
                return new Position($this->row, ($this->col - 1 + $cols) % $cols);
            default:
                return clone $this;
        }
    }

    /**
     * Calculate squared distance with toroidal wrapping
     */
    public function distance2(Position $other, int $rows, int $cols): int {
        $dr = abs($this->row - $other->row);
        $dc = abs($this->col - $other->col);
        $dr = min($dr, $rows - $dr);
        $dc = min($dc, $cols - $dc);
        return $dr * $dr + $dc * $dc;
    }
}

/**
 * Game configuration
 */
class GameConfig {
    public int $rows;
    public int $cols;
    public int $maxTurns;
    public int $visionRadius2;
    public int $attackRadius2;
    public int $spawnCost;
    public int $energyInterval;

    public static function fromArray(array $data): self {
        $config = new self();
        $config->rows = $data['rows'];
        $config->cols = $data['cols'];
        $config->maxTurns = $data['max_turns'];
        $config->visionRadius2 = $data['vision_radius2'];
        $config->attackRadius2 = $data['attack_radius2'];
        $config->spawnCost = $data['spawn_cost'];
        $config->energyInterval = $data['energy_interval'];
        return $config;
    }
}

/**
 * Player info
 */
class PlayerInfo {
    public int $id;
    public int $energy;
    public int $score;

    public static function fromArray(array $data): self {
        $info = new self();
        $info->id = $data['id'];
        $info->energy = $data['energy'];
        $info->score = $data['score'];
        return $info;
    }
}

/**
 * Visible bot
 */
class VisibleBot {
    public Position $position;
    public int $owner;

    public static function fromArray(array $data): self {
        $bot = new self();
        $bot->position = Position::fromArray($data['position']);
        $bot->owner = $data['owner'];
        return $bot;
    }
}

/**
 * Visible core
 */
class VisibleCore {
    public Position $position;
    public int $owner;
    public bool $active;

    public static function fromArray(array $data): self {
        $core = new self();
        $core->position = Position::fromArray($data['position']);
        $core->owner = $data['owner'];
        $core->active = $data['active'];
        return $core;
    }
}

/**
 * Fog-filtered game state
 */
class GameState {
    public string $matchId;
    public int $turn;
    public GameConfig $config;
    public PlayerInfo $you;
    /** @var VisibleBot[] */
    public array $bots = [];
    /** @var Position[] */
    public array $energy = [];
    /** @var VisibleCore[] */
    public array $cores = [];
    /** @var Position[] */
    public array $walls = [];
    /** @var VisibleBot[] */
    public array $dead = [];

    public static function fromArray(array $data): self {
        $state = new self();
        $state->matchId = $data['match_id'];
        $state->turn = $data['turn'];
        $state->config = GameConfig::fromArray($data['config']);
        $state->you = PlayerInfo::fromArray($data['you']);

        foreach ($data['bots'] ?? [] as $bot) {
            $state->bots[] = VisibleBot::fromArray($bot);
        }

        foreach ($data['energy'] ?? [] as $pos) {
            $state->energy[] = Position::fromArray($pos);
        }

        foreach ($data['cores'] ?? [] as $core) {
            $state->cores[] = VisibleCore::fromArray($core);
        }

        foreach ($data['walls'] ?? [] as $pos) {
            $state->walls[] = Position::fromArray($pos);
        }

        foreach ($data['dead'] ?? [] as $bot) {
            $state->dead[] = VisibleBot::fromArray($bot);
        }

        return $state;
    }
}

/**
 * A single move command
 */
class Move {
    public Position $position;
    public string $direction;

    public function __construct(Position $position, string $direction) {
        $this->position = $position;
        $this->direction = $direction;
    }

    public function toArray(): array {
        return [
            'position' => $this->position->toArray(),
            'direction' => $this->direction
        ];
    }
}

File: bots/assassin/src/game.rs
//! Game state types for AI Code Battle protocol.

use serde::{Deserialize, Serialize};

/// Position on the grid
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Position {
    pub row: i32,
    pub col: i32,
}

/// Game configuration
#[derive(Debug, Clone, Deserialize)]
pub struct GameConfig {
    pub rows: u32,
    pub cols: u32,
    pub max_turns: u32,
    pub vision_radius2: u32,
    pub attack_radius2: u32,
    pub spawn_cost: u32,
    pub energy_interval: u32,
}

/// Player info
#[derive(Debug, Clone, Deserialize)]
pub struct PlayerInfo {
    pub id: u32,
    pub energy: u32,
    pub score: u32,
}

/// Visible bot
#[derive(Debug, Clone, Deserialize)]
pub struct VisibleBot {
    pub position: Position,
    pub owner: u32,
}

/// Visible core
#[derive(Debug, Clone, Deserialize)]
pub struct VisibleCore {
    pub position: Position,
    pub owner: u32,
    pub active: bool,
}

/// Fog-filtered game state visible to this bot
#[derive(Debug, Clone, Deserialize)]
pub struct GameState {
    pub match_id: String,
    pub turn: u32,
    pub config: GameConfig,
    pub you: PlayerInfo,
    #[serde(default)]
    pub bots: Vec<VisibleBot>,
    #[serde(default)]
    pub energy: Vec<Position>,
    #[serde(default)]
    pub cores: Vec<VisibleCore>,
    #[serde(default)]
    pub walls: Vec<Position>,
    #[serde(default)]
    pub dead: Vec<VisibleBot>,
}

/// Movement direction
#[derive(Debug, Clone, Copy, Serialize)]
pub enum Direction {
    #[serde(rename = "N")]
    N,
    #[serde(rename = "E")]
    E,
    #[serde(rename = "S")]
    S,
    #[serde(rename = "W")]
    W,
}

/// A single move command
#[derive(Debug, Clone, Serialize)]
pub struct Move {
    pub position: Position,
    pub direction: Direction,
}

/// Response containing moves
#[derive(Debug, Clone, Serialize)]
pub struct MoveResponse {
    pub moves: Vec<Move>,
}

impl Direction {
    /// All directions in order: N, E, S, W
    pub fn all() -> [Direction; 4] {
        [Direction::N, Direction::E, Direction::S, Direction::W]
    }
}

impl Position {
    /// Move in a direction, wrapping around the toroidal grid
    pub fn move_toward(&self, dir: Direction, rows: i32, cols: i32) -> Position {
        match dir {
            Direction::N => Position {
                row: (self.row - 1 + rows) % rows,
                col: self.col,
            },
            Direction::E => Position {
                row: self.row,
                col: (self.col + 1) % cols,
            },
            Direction::S => Position {
                row: (self.row + 1) % rows,
                col: self.col,
            },
            Direction::W => Position {
                row: self.row,
                col: (self.col - 1 + cols) % cols,
            },
        }
    }

    /// Calculate squared distance with toroidal wrapping
    pub fn distance2(&self, other: &Position, rows: i32, cols: i32) -> u32 {
        let dr = (self.row - other.row).abs();
        let dc = (self.col - other.col).abs();
        let dr = dr.min(rows - dr);
        let dc = dc.min(cols - dc);
        (dr * dr + dc * dc) as u32
    }
}
